drivers/serial/uart_nrfx_uarte.c

```diff
@@ -0,0 +1,651 @@
+
+	int err;
+	struct device *gpio_dev;
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+
+	gpio_dev = device_get_binding(CONFIG_GPIO_NRF5_P0_DEV_NAME);
+
+	__ASSERT(gpio_dev,
+		 "UART init failed. Cannot find %s",
+		 CONFIG_GPIO_NRF5_P0_DEV_NAME);
+
+	gpio_pin_write(gpio_dev, config->pseltxd, 1);
+	gpio_pin_configure(gpio_dev,
+			   config->pseltxd,
+			   GPIO_DIR_OUT);
+	gpio_pin_configure(gpio_dev,
+			   config->pselrxd,
+			   GPIO_DIR_IN);
+
+	nrf_uarte_txrx_pins_set(uarte,
+				config->pseltxd,
+				config->pselrxd);
+
+	if (config->hwfc == NRF_UARTE_HWFC_ENABLED) {
+		gpio_pin_write(gpio_dev, config->pselrts, 1);
+		gpio_pin_configure(gpio_dev,
+				   config->pselrts,
+				   GPIO_DIR_OUT);
+		gpio_pin_configure(gpio_dev,
+				   config->pselcts,
+				   GPIO_DIR_IN);
+		nrf_uarte_hwfc_pins_set(uarte,
+					config->pselrts,
+					config->pselcts);
+	}
+
+	/* Set low control and parity bit */
+	nrf_uarte_configure(uarte,
+			    config->parity,
+			    config->hwfc);
+
+	/* Set baud rate */
+	err = baudrate_set(dev, config->baudrate);
+	if (err) {
+		return err;
+	}
+
+	/* Enable receiver and transmitter */
+	nrf_uarte_enable(uarte);
+
+	nrf_uarte_event_clear(uarte, NRF_UARTE_EVENT_ENDRX);
+
+	nrf_uarte_rx_buffer_set(uarte, &data->rx_data, 1);
+	nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTRX);
+
+#if CONFIG_UART_INTERRUPT_DRIVEN
+	if (interrupts_active) {
+		u8_t dummy_byte = 0;
+
+		irq_enable(NRFX_IRQ_NUMBER_GET(uarte));
+
+		/* Set ENDTX event by sending empty frame */
+		nrf_uarte_tx_buffer_set(uarte, &dummy_byte, 0);
+		nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTTX);
+
+		/* switch off transmitter to save an energy */
+		nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STOPTX);
+	}
+#endif
+
+	return 0;
+}
+
+#define UART_NRF_UARTE_DEVICE(idx)					      \
+	static struct uarte_nrfx_data uarte_##idx##_data;		      \
+	static const struct uarte_nrfx_config uarte_##idx##_config = {	      \
+		.max_chunk_len = (1 << UARTE##idx##_EASYDMA_MAXCNT_SIZE) - 1, \
+		.uarte.p_reg = (NRF_UARTE_Type *)NRF_UARTE##idx		      \
+	};								      \
+	static int uarte_##idx##_init(struct device *dev)		      \
+	{								      \
+		const struct uarte_init_config init_config = {		      \
+		.pseltxd = CONFIG_UART_##idx##_NRF_TX_PIN,		      \
+		.pselrxd = CONFIG_UART_##idx##_NRF_RX_PIN,		      \
+		.pselcts = CONFIG_UART_##idx##_NRF_CTS_PIN,		      \
+		.pselrts = CONFIG_UART_##idx##_NRF_RTS_PIN,		      \
+		.hwfc = CONFIG_UART_##idx##_NRF_FLOW_CONTROL,		      \
+		.parity = CONFIG_UART_##idx##_NRF_PARITY_BIT,		      \
+		.baudrate = CONFIG_UART_##idx##_BAUD_RATE		      \
+		};							      \
+		if (CONFIG_UART_##idx##_INTERRUPT_DRIVEN) {		      \
+			IRQ_CONNECT(NRFX_IRQ_NUMBER_GET(NRF_UARTE##idx),      \
+				    CONFIG_UART_##idx##_IRQ_PRI,              \
+				    uarte_nrfx_isr,                           \
+				    DEVICE_GET(uart_nrfx_uarte##idx),         \
```

Comment:
Since you don't use anything else from `dev` in the IRQ handler, you could pass `&uarte_##idx##_data` here and simplify the code (you'll get rid of `DEVICE_DECLARE`, too).

Reply:
Cannot do it as callback function requires devive structure.

Reply:
Right. My mistake.